ITR/data/base_providers.py

```diff
@@ -567,40 +640,66 @@ def _standardize(self, intensities: pd.DataFrame) -> pd.DataFrame:
+        units = historic_intensities.apply(lambda x: x[x.first_valid_index()].u
+                                           if x.first_valid_index() and isinstance(x[x.first_valid_index()], pint.Quantity) else None)
+        # FIXME: we already remove all-NA rows before this function is called.  Can any non-unit data leak to here?
+        na_units = units.isna()
+        if na_units.any():
+            logger.warning(f"Winsorization dropping {na_units[na_units].index.to_list()} due to null unit information")
+            historic_intensities = historic_intensities[~units_na]
+        if ITR.HAS_UNCERTAINTIES:
+            nominal_intensities = historic_intensities.apply(lambda x: ITR.nominal_values(x.map(lambda y: y.m if isinstance(y, pint.Quantity) else np.nan)))
+            uncertain_intensities = historic_intensities.apply(lambda x: ITR.std_devs(x.map(lambda y: y.m if isinstance(y, pint.Quantity) else 0)))
+        else:
+            nominal_intensities = historic_intensities.apply(lambda x: x.map(lambda y: y.m if isinstance(y, pint.Quantity) else np.nan))
+        # See https://github.com/hgrecco/pint-pandas/issues/114
+        lower=nominal_intensities.quantile(q=self.projection_controls.LOWER_PERCENTILE, axis='index', numeric_only=False)
+        upper=nominal_intensities.quantile(q=self.projection_controls.UPPER_PERCENTILE, axis='index', numeric_only=False)
+        winsorized: pd.DataFrame = nominal_intensities.clip(
+            lower=lower,
+            upper=upper,
+            axis='columns'
+        )
+        if ITR.HAS_UNCERTAINTIES:
+            # FIXME: the clipping process can properly introduce uncertainties.  The low and high values that are clipped could be
+            # replaced by the clipped values +/- the lower and upper percentile values respectively.
+            if uncertain_intensities.values.sum() != 0:
+                uwinsorized = winsorized.apply(lambda x: PA_(ITR.uarray(x.values.data, uncertain_intensities[x.name].values), dtype=units[x.name]))
+                return uwinsorized
+        # FIXME: If we have S1, S2, and S1S2 intensities, should we treat winsorized(S1)+winsorized(S2) as winsorized(S1S2)?
+        # FIXME: If we have S1S2 (or S1 and S2) and S3 and S1S23 intensities, should we treat winsorized(S1S2)+winsorized(S3) as winsorized(S1S2S3)?
+        winsorized_and_unitized = winsorized.apply(lambda x: PA_(x.values.data, dtype=units[x.name]))
+        return winsorized_and_unitized
 
     def _interpolate(self, historic_intensities: pd.DataFrame) -> pd.DataFrame:
         # Interpolate NaNs surrounded by values, and extrapolate NaNs with last known value
-        interpolated = historic_intensities.copy()
-        for col in interpolated.columns:
-            if interpolated[col].isnull().all():
+        df = historic_intensities.copy()
+        for col in df.columns:
+            pa = PA_._from_sequence(df[col])
+            if ITR.isnan(pa.data).all():
```

Comment:
originally, here was `isnull`.
when we replace it with `isnan`, what's the benefit?
was original check broken?

Reply:
ITR.isnan is intended to be consistent across both floats and uncertainties.  isnull won't see nans with uncertainties.

Reply:
Do you want to write some unit tests for interpolation, or should I?

Reply:
> ITR.isnan is intended to be consistent across both floats and uncertainties.

got it, it answers my question!

> Do you want to write some unit tests for interpolation

your answer already clarifies my question